src/military/military.cpp

```diff
@@ -3940,7 +3947,7 @@ void army_arrives_in_province(sys::state& state, dcon::army_id a, dcon::province
 
 	state.world.army_set_location_from_army_location(a, p);
 	auto regs = state.world.army_get_army_membership(a);
-	if(!state.world.army_get_black_flag(a) && !state.world.army_get_is_retreating(a) && regs.begin() != regs.end()) {
+	if(army_can_join_land_battle(state, a) && regs.begin() != regs.end()) {
```

Comment:
note that this will be slower: you are now checking for more things, one of which should have been eliminated by the assert (and the other shouldn't be relevant because this function isn't called for armies being transported -- unless it is called while they are moving off the boats but before that relationship has been broken, in which case this change will be a bug)